### DISCUSSION TITLE
Activate conda environment before python executable runs

### DISCUSSION THREAD

**Discussion by scottwsides** [pyinstaller/pyinstaller#6168] [PyInstaller]
I have a conda environment with all my dependencies set. I dont want pyinstaller to package these since its complicated and pyinstaller is not getting them correct. I simply need my PyQt app (the one listed in the Analysis class) to be running in a conda environment I've created.. then all imports will be found. I've tried lots of ideas to rig this ... but nothing works. Setting a path for my conda environment in 'pathex' doesnt work either. All attempts to bootstrap this by running scripts in my python executable dont work. Is there a solution?

**Comment by eric15342335** [comment]
> I dont want pyinstaller to package these since its complicated and pyinstaller is not getting them correct.

What is the error you have encountered?

**Reply by scottwsides** [reply]
When I try to run 'open XXXXX.app' the applications starts up and then closes, presumably because it cannot find the extra dependencies.

**Comment by rokm** [comment]
PyInstaller-frozen executable runs your frozen script in embedded python interpreter that is self-contained, i.e., limited to what was collected into frozen bundle during the build phase. If your PyQt app needs other dependencies from the Anaconda environment to work, they need to be present during the build (so you need to build with your conda environment activated). 

I suppose you could try to emulate the activation of conda environment at the start of your frozen python script: extend `sys.path`, set DLL directories (and/or set `PATH`), set other environment variables, etc. But if you plan to go down that route, why bother with PyInstaller at all? Your semi-frozen application will need a full-blown conda environment installed, so you might as well run it all from source...

**Reply by scottwsides** [reply]
I do run pyinstaller build with the conda environment activated. Some of my dependency directories look like they got partially setup... but not all the source code is there. I need pyinstaller because it seemed (at the time) like the best cross-platform way of making a XXXXX.app/ (for OS X) and XXXXXX.exe for windows.  But I take your point.. part of what pyinstaller is doing is packaging dependencies and I've got conda for that. So what's the best way of making the clickable 'app' and 'exe' files?

**Reply by rokm** [reply]
> but not all the source code is there

The source code should not be there at all (except in some special cases), because PyInstaller collects byte-compiled .pyc modules (which are stored in executable-embedded archive, and not as individual files - unless you use `--debug noarchive˙).

**Reply by rokm** [reply]
But from what you said, you should probably first try making POSIX (console) frozen application on macOS, and only once you get that up and running, move on to creating the .app bundles ("windowed" mode). Otherwise, you'll have a difficult time figuring out what part is going wrong (i.e., dependencies or some related error, or something bundle-related). The unhandled exceptions in .app bundles should now be logged in macOS' `Console.app`, but that seems to work only if with python versions that target older macOS SDK version (so if your version targets a more recent version, you should implement top-level exception handling and log the exceptions somewhere, for example to file. This will give you an idea of what is going on....)

**Comment by scottwsides** [comment]
I've got the application running from the raw executable I want to start of course. From your suggestion... I think pyinstaller is a lot of overhead for what I actually need. When I first started developing this... I had the PyQt dependences in a local Framework python that I needed packaged. Now these are in the same conda environment that I distribute with conda build and constructor tools anyway. The step of just making an app bundle is too simple probably to mess with pyinstlaler.

**Reply by rokm** [reply]
Indeed. You could probably create an .app bundle with bash script for the entry-point "executable", which in turn launches python interpreter from your anaconda environment and runs your program. And similarly, you could launch python and your script via a batch file on Windows (which would make it "clickable"). There's little point in complicating your life with PyInstaller if you are not interested in freezing the application in the first place...

**Reply by scottwsides** [reply]
Exactly! Thanks for breaking the spell. There were also historical reasons why I was using pyinstaller before I switched to using conda to do heavy lifting. The pyinstaller dependency was really a vestige.
